keras_maskrcnn/preprocessing/csv_generator.py

```diff
@@ -0,0 +1,189 @@
+        result[class_name] = class_id
+    return result
+
+
+def _read_annotations(csv_reader, classes):
+    result = {}
+    for line, row in enumerate(csv_reader):
+        try:
+            img_file, x1, y1, x2, y2, class_name, mask_path = row
+        except ValueError:
+            raise_from(ValueError('line {}: format should be \'img_file,x1,y1,x2,y2,class_name,mask_path\' or \'img_file,,,,,\''.format(line)), None)
+
+        if img_file not in result:
+            result[img_file] = []
+
+        #if mask_path not in result:
+        #    result[mask_path] = []
+
+        # If a row contains only an image path, it's an image without annotations.
+        if (x1, y1, x2, y2, class_name) == ('', '', '', '', ''):
+            continue
+
+        x1 = _parse(x1, int, 'line {}: malformed x1: {{}}'.format(line))
+        y1 = _parse(y1, int, 'line {}: malformed y1: {{}}'.format(line))
+        x2 = _parse(x2, int, 'line {}: malformed x2: {{}}'.format(line))
+        y2 = _parse(y2, int, 'line {}: malformed y2: {{}}'.format(line))
+
+        # Check that the bounding box is valid.
+        if x2 <= x1:
+            raise ValueError('line {}: x2 ({}) must be higher than x1 ({})'.format(line, x2, x1))
+        if y2 <= y1:
+            raise ValueError('line {}: y2 ({}) must be higher than y1 ({})'.format(line, y2, y1))
+
+        # check if the current class name is correctly present
+        if class_name not in classes:
+            raise ValueError('line {}: unknown class name: \'{}\' (classes: {})'.format(line, class_name, classes))
+
+        result[img_file].append({'x1': x1, 'x2': x2, 'y1': y1, 'y2': y2, 'class': class_name, 'mask_path': mask_path})
+    return result
+
+
+def _open_for_csv(path):
+    """
+    Open a file with flags suitable for csv.reader.
+
+    This is different for python2 it means with mode 'rb',
+    for python3 this means 'r' with "universal newlines".
+    """
+    if sys.version_info[0] < 3:
+        return open(path, 'rb')
+    else:
+        return open(path, 'r', newline='')
+
+
+class CSVGenerator(Generator):
+    def __init__(
+        self,
+        csv_data_file,
+        csv_class_file,
+        base_dir=None,
+        **kwargs
+    ):
+        self.image_names = []
+        self.image_data  = {}
+        self.base_dir    = base_dir
+
+        # Take base_dir from annotations file if not explicitly specified.
+        if self.base_dir is None:
+            self.base_dir = os.path.dirname(csv_data_file)
+
+        # parse the provided class file
+        try:
+            with _open_for_csv(csv_class_file) as file:
+                self.classes = _read_classes(csv.reader(file, delimiter=','))
+        except ValueError as e:
+            raise_from(ValueError('invalid CSV class file: {}: {}'.format(csv_class_file, e)), None)
+
+        self.labels = {}
+        for key, value in self.classes.items():
+            self.labels[value] = key
+
+        # csv with img_path, x1, y1, x2, y2, class_name
```

Comment:
Not correct anymore.

Reply:
well, it was 6 days ago when I sent the pull request

Reply:
Hmm but six days ago this also included the mask_path?

Reply:
wait, but are you talking about the comment?

Reply:
yeah